plugins/login-resources/src/components/Form.svelte

```diff
@@ -66,7 +67,19 @@
 
   $: $themeStore.language && validate($themeStore.language)
 
+  const validationMutex = new Mutex()
```

Comment:
Hi, thanks for figure this out.

I think we need a bit more generic solution, since we potentially have more cases like this one, there we should ignore results or be sore operations are completed one but one, but proposed solution looks a bit more complicated.

We usually use local Promise variable and add await inside update method, but missed in this particular place, if you have some use-full generic idea it would be great.

Thanks.

Reply:
Thanks for looking into this, I am happy to work on improving the proposed solution, but I need a few clarifications:

> 1. proposed solution looks a bit more complicated

Is it particularly about  _implementation_ of Mutex or its _usage_?

> 2. there we should ignore results or be sore operations are completed one but one

I didn't fully understand this part (probably there were some typos), could you rephrase?

> 3. We usually use local Promise variable and add await inside update method

Could you link an example?

---

Also, what would be the right package/place in this repository to put a generic utility like Mutex into?

Reply:
Hi @vlad-timofeev sorry for delay with response.

1. Usage one.

I suppose we should evolve your idea to something like this.

```typescript
# Some there in platform
  export function oneByOne<T extends (...args: Parameters<T>) => Promise<void>> (t: T): (...args: Parameters<T>) => Promise<void> {
    let previousCall: Promise<any> | undefined
    return async function (...args: Parameters<T>): Promise<void> {
      if (previousCall !== undefined) {
        await previousCall
      }
      previousCall = t(...args)
      return await previousCall
    }
  }
# A component code
  let a = 20
  const myUpdate = oneByOne(async (x: number) => {
    console.log('called with', x)
    await new Promise<void>((resolve) => {
      setTimeout(resolve, x)
    })
    a = x
    console.log('returned with', x)
  })
  myUpdate(200)
  myUpdate(100)
```
So queueCalls would hold a promise of previous execution and before calling myUpdate will wait for previous to complete, so if variables are changing in different order and we have some asyncs inside update function, we will use results of last one arguments. Because assume `something` for some previous arguments could work longer.

Reply:
Nope, this variant is not working as expected, this one is correct.

```typescript
 function reduceCalls<T extends (...args: Parameters<T>) => Promise<void>> (t: T): (...args: Parameters<T>) => Promise<void> {
    let nextCalls: (() => Promise<void>)[] = []
    return async function (...args: Parameters<T>): Promise<void> {
      const myOp = async (): Promise<void> => {
        await t(...args)
        nextCalls.splice(0, 1)
        const last = nextCalls[nextCalls.length - 1]
        nextCalls = [last]
        if (last !== undefined) {
          void last()
        }
      }
      nextCalls.push(myOp)
      if (nextCalls.length === 1) {
        await myOp()
      }
    }
  }

  let a = 20
  const myUpdate = reduceCalls(async (x: number) => {
    console.log('called with', x)
    await new Promise<void>((resolve) => {
      setTimeout(resolve, x)
    })
    a = x
    console.log('returned with', x)
  })
  myUpdate(500)
  myUpdate(100)
  myUpdate(50)
```

Reply:
Thanks for that. I will re-work this PR and push a new revision asap.

Reply:
@haiodo , I tested your variants locally , the second one printed a different result from what I expected though:

```
[user@huly] ➜  ~/workshops/promise_test npx tsx from_mr_ace.ts                                                                                                                                                                 INSERT
called with 500
returned with 500
called with 50
returned with 50
[user@huly] ➜  ~/workshops/promise_test                                                                                                                                                                                       
```

I used the function-wrapper interface that you proposed, and encapsulated the existing Mutex implementation inside. I appended some code to `mutex.ts` file in the 2nd commit to expose this interface and as a result it is just as simple to use as in your examples. I also added tests to verify that this functionality works as expected.